### ISSUE TITLE
[Feature Request] Add an option to run GPU tests only, and skip all CPU tests

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#44159]
## 🚀 Feature

Add an option to run GPU tests only, and skip all CPU tests

For example, we can have an environment variable `PYTORCH_TEST_GPU_ONLY=1`.

## Motivation

Usually, when people only want to test GPU functionalities, they can skip CPU tests to save plenty of times. 

## Pitch

For example, some pytorch tests look like (assuming pytorch built with GPU)
```python
def test_something(self, device):
    # do something
```
At runtime, two tests `test_something_cpu` and `test_something_gpu` will be generated. The goal is to skip the `test_something_cpu` test and only run `test_something_gpu`. 

For example, we can do it like this
```bash
PYTORCH_TEST_GPU_ONLY=1 python test/run_test.py
```

## Alternatives

N/A

## How to implement 

If I understand the code correctly, this feature can be easily implemented like this

1. Add `TEST_GPU_ONLY = os.getenv('PYTORCH_TEST_GPU_ONLY', '0') == '1' ` here
https://github.com/pytorch/pytorch/blob/665feda15bc45d0f50326596ecde6f2d96ac6644/torch/testing/_internal/common_utils.py#L346-L347

2. Add something like
```python
    if TEST_GPU_ONLY:
        fullname = self.id().lower()
        is_cuda_test = 'gpu' in fullname or 'cuda' in fullname or 'cuda' in torch.tensor([]).device.type
        if not is_cuda_test:
            raise unittest.SkipTest("Test is probably not a GPU test. We disabled it with PYTORCH_TEST_GPU_ONLY")
```
here in `setup` of `class common_utils.TestCase`
https://github.com/pytorch/pytorch/blob/665feda15bc45d0f50326596ecde6f2d96ac6644/torch/testing/_internal/common_utils.py#L825-L833

## Additional context

N/A 

cc @ptrblck @ngimel @mruberry

**Comment by kshitij12345** [comment]
@xwang233 
How about
```shell
pytest test/test_torch.py -k _cuda_
```
Should do the same thing

**Comment by mruberry** [comment]
@kshitij12345's suggestion is very practical and likely to do most of you'd like. We could also build a more formal mechanism. For example, in common_device_type we could instantiate only the cpu or cuda (or xla...) device types if the appropriate flag was set. Not all of our test suite uses the device generic test framework, however. And, ironically, test_cuda.py would probably need to be made generic for this approach to work, too. There might also need to be a way to run only the test suites that are generic. (Maybe the test runner could query to see if the test class is device generic or not?)

We would definitely accept PRs making more of the test suite device generic. I think we would also accept a PR that limited the test suite to running device generic tests for only one device.

**Comment by walterddr** [comment]
Hmm. I don't think there's a per test method level filtering. We still leverage `-k` option in unittest/pytest to do the test method name pattern filter. 
Just to clarify. isn't #55174 a super set of this issue?

**Comment by walterddr** [comment]
Also i dont know if this is related. I think there's 2 flags being reworked in https://github.com/pytorch/pytorch/pull/55753 that allows one to control what device_type to generate test cases. 
Thus the mechanism to control tests generated via device-generic framework is there.

**Comment by mruberry** [comment]
> Hmm. I don't think there's a per test method level filtering. We still leverage `-k` option in unittest/pytest to do the test method name pattern filter.
> Just to clarify. isn't #55174 a super set of this issue?
> also is there an issue tracker / checklist for moving more tests to device generic framework?

I don't think we have a tracker, but if someone is seriously pursuing issues like this it would be helpful.
